src/devices/bus/sg1000_exp/sp400.cpp

```diff
@@ -0,0 +1,249 @@
+{
+	map(0x0000, 0x0fff).rom();
+}
+
+void sp400_printer_device::dser_p1_w(uint8_t data)
+{
+	m_dserdataout = data;
+}
+
+void sp400_printer_device::dser_p2_w(uint8_t data)
+{
+	m_busy = BIT(data, 5);
+	m_dserstrobe = BIT(data, 6);
+	m_motcpu->set_input_line(M6805_IRQ_LINE, m_dserstrobe);
+}
+
+uint8_t sp400_printer_device::dser_p2_r()
+{
+	return
+		(m_motbusy << 7) |
+		(m_dserstrobe << 6) |
+		(m_busy << 5) |
+		(m_data << 4) |
+		(1 << 3) |
+		(1 << 2) |
+		(1 << 1) |
+		(1 << 0);
+}
+
+uint8_t sp400_printer_device::mot_pa_r()
+{
+	return
+		(0 << 0) |
+		(0 << 1) |
+		(m_buttons->read() << 2) |  // bits 2..4
+		(0 << 5) |
+		(0 << 6) |
+		(1 << 7);
+}
+
+void sp400_printer_device::mot_pb_w(uint8_t data)
+{
+	// cr stepper data in 0..3
+	// pf stepper data in 4..7
+}
+
+uint8_t sp400_printer_device::mot_pc_r()
+{
+	return
+		(m_motPenUp   << 0) |
+		(m_motPenDown << 1) |
+		(1 << 2) |
+		(pen_sensor() << 3) |
+		(m_motbusy    << 4) |
+		(1 << 5) | (1 << 6) | (1 << 7); //unconfirmed
+}
+
+void sp400_printer_device::mot_pc_w(uint8_t data)
+{
+	m_motPenUp   = BIT(data, 0);
+	m_motPenDown = BIT(data, 1);
+	m_motbusy    = BIT(data, 4);
+	update_pen_state(data);
+}
+
+uint8_t sp400_printer_device::mot_pd_r()
+{
+	return sp400_printer_device::m_dserdataout;
+}
+
+void sp400_printer_device::update_pen_state(u8 data)
+{
+	if (!BIT(data, 0)) m_pendown = 0;
+	if (!BIT(data, 1)) m_pendown = 1;
+}
+
+void sp400_printer_device::device_add_mconfig(machine_config &config)
+{
+	I8035(config, m_dsercpu, 4_MHz_XTAL);
+	m_dsercpu->p1_out_cb().set(FUNC(sp400_printer_device::dser_p1_w));
+	m_dsercpu->p2_in_cb().set(FUNC(sp400_printer_device::dser_p2_r));
+	m_dsercpu->p2_out_cb().set(FUNC(sp400_printer_device::dser_p2_w));
+	m_dsercpu->set_addrmap(AS_PROGRAM, &sp400_printer_device::dser_map);
+
+	M6805U3(config, m_motcpu, 4_MHz_XTAL);
+	m_motcpu->porta_r().set(FUNC(sp400_printer_device::mot_pa_r));
+	m_motcpu->portb_w().set(FUNC(sp400_printer_device::mot_pb_w));
+	m_motcpu->portc_r().set(FUNC(sp400_printer_device::mot_pc_r));
+	m_motcpu->portc_w().set(FUNC(sp400_printer_device::mot_pc_w));
+	m_motcpu->portd_r().set(FUNC(sp400_printer_device::mot_pd_r));
+}
+
+ROM_START( sp400 )
+	ROM_REGION( 0x1000, "dsercpu", 0 )
+	ROM_LOAD( "sp400_8035.bin", 0x0000, 0x1000, CRC(0eb48272) SHA1(08a2727f1592f5d2ecb2e368126ad7bfc5d3c270) )
+
+	ROM_REGION(0x1000, "motcpu", 0 )
+	ROM_LOAD( "sp400_hd6805v1.bin", 0x0000, 0x1000, CRC(aa073745) SHA1(65016f3b022af30cc6b084af1e43b29168721a60) )
+ROM_END
+
+const tiny_rom_entry * sp400_printer_device::device_rom_region() const
+{
+	return ROM_NAME( sp400 );
+}
+
+static INPUT_PORTS_START( sp400 )
+	PORT_START("BUTTONS")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Color Select") PORT_CODE(KEYCODE_2_PAD)
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Line Feed")    PORT_CODE(KEYCODE_1_PAD)
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Pen Change")   PORT_CODE(KEYCODE_3_PAD)
+INPUT_PORTS_END
```

Comment:
You don’t actually need `static` to indicate that you don’t want something visible from outside the translation unit from inside an anonymous namespace.

Also, these probably shouldn’t be `IPT_KEYBOARD`.  Maybe `IPT_KEYPAD` or `IPT_OTHER`?